src/utils/random-data.js

```diff
@@ -176,54 +177,38 @@ class Pipeline {
    * @returns {array} Edge objects
    */
   generateEdges() {
+    const nodesByRank = [...this.nodes];
+    nodesByRank.sort((a, b) => a.rank - b.rank);
+
     const edges = [];
-    const dataNodes = this.nodes.filter(node => node.type === 'data');
-    const taskNodes = this.nodes.filter(node => node.type !== 'data');
-    const remainingEdgeCount = arrayToObject(
-      dataNodes.map(node => node.id),
-      () => MAX_CONNECTED_NODES
-    );
+    const maxEdgeCount = randomNumberBetween(MIN_EDGE_COUNT, MAX_EDGE_COUNT);
+
+    for (let i = 0; i < maxEdgeCount; i += 1) {
+      const sourceIndex = randomIndex(nodesByRank.length);
+      const source = nodesByRank[sourceIndex];
 
-    taskNodes.forEach(node => {
-      const ancestors = dataNodes.filter(
-        d => d.rank < node.rank && remainingEdgeCount[d.id]
+      const nextRankFirstIndex = nodesByRank.findIndex(
+        (node, index) => index > sourceIndex && node.rank > source.rank
       );
-      this.getRandomNodes(ancestors).forEach(source => {
-        edges.push({
-          source: source.id,
-          target: node.id
-        });
-        remainingEdgeCount[source.id]--;
-      });
 
-      const descendants = dataNodes.filter(
-        d => d.rank > node.rank && remainingEdgeCount[d.id]
+      if (nextRankFirstIndex === -1) continue;
+
+      const successorCount = nodesByRank.length - nextRankFirstIndex - 1;
+      const randomSuccessor = Math.round(
+        Math.min(1 / Math.random(), successorCount)
```

Comment:
This usage of `Math.random` won't work with the data seeding, so it might be worth splitting it out into the utility functions

Reply:
Good catch, I've updated this and on another line it was still in use too.